pkg/v1/tarball/write.go

```diff
@@ -31,115 +32,198 @@ import (
-// for the written images as Manifest.`If the tar.Writer is nil, will just return
-// the Manifest.
-func processImages(refToImage map[name.Reference]v1.Image, tf *tar.Writer) (Manifest, error) {
+// writeImagesToTar writes the images to the tarball
+func writeImagesToTar(refToImage map[name.Reference]v1.Image, m []byte, size int64, w io.Writer, o *writeOptions) (err error) {
+	if w == nil {
+		return errors.New("must pass valid writer")
+	}
 	imageToTags := dedupRefToImage(refToImage)
-	var m Manifest
+
+	tw := w
+
+	// we only calculate the sizes and use a progressWriter if we were provided
+	// an option with a progress channel
+	if o != nil && o.updates != nil {
+		tw = &progressWriter{
+			w:       w,
+			updates: o.updates,
+			size:    size,
+		}
+	}
+
+	tf := tar.NewWriter(tw)
+	defer tf.Close()
 
 	seenLayerDigests := make(map[string]struct{})
 
-	for img, tags := range imageToTags {
+	for img := range imageToTags {
 		// Write the config.
 		cfgName, err := img.ConfigName()
 		if err != nil {
-			return nil, err
+			return err
+		}
+		cfgBlob, err := img.RawConfigFile()
+		if err != nil {
+			return err
+		}
+		if err := writeTarEntry(tf, cfgName.String(), bytes.NewReader(cfgBlob), int64(len(cfgBlob))); err != nil {
+			return err
+		}
+
+		// Write the layers.
+		layers, err := img.Layers()
+		if err != nil {
+			return err
 		}
-		if tf != nil {
-			cfgBlob, err := img.RawConfigFile()
+		layerFiles := make([]string, len(layers))
+		for i, l := range layers {
+			d, err := l.Digest()
+			if err != nil {
+				return err
+			}
+			// Munge the file name to appease ancient technology.
+			//
+			// tar assumes anything with a colon is a remote tape drive:
+			// https://www.gnu.org/software/tar/manual/html_section/tar_45.html
+			// Drop the algorithm prefix, e.g. "sha256:"
+			hex := d.Hex
+
+			// gunzip expects certain file extensions:
+			// https://www.gnu.org/software/gzip/manual/html_node/Overview.html
+			layerFiles[i] = fmt.Sprintf("%s.tar.gz", hex)
+
+			if _, ok := seenLayerDigests[hex]; ok {
+				continue
+			}
+			seenLayerDigests[hex] = struct{}{}
+
+			r, err := l.Compressed()
+			if err != nil {
+				return err
+			}
+			blobSize, err := l.Size()
 			if err != nil {
-				return nil, err
+				return err
 			}
-			if err := writeTarEntry(tf, cfgName.String(), bytes.NewReader(cfgBlob), int64(len(cfgBlob))); err != nil {
-				return nil, err
+
+			if err := writeTarEntry(tf, layerFiles[i], r, blobSize); err != nil {
+				return err
 			}
 		}
+	}
+	if err := writeTarEntry(tf, "manifest.json", bytes.NewReader(m), int64(len(m))); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// calculateManifest calculates the manifest and optionally the size of the tar file
+func calculateManifest(refToImage map[name.Reference]v1.Image) (m Manifest, err error) {
+	imageToTags := dedupRefToImage(refToImage)
+
+	for img, tags := range imageToTags {
+		cfgName, err := img.ConfigName()
+		if err != nil {
+			return m, err
```

Comment:
Sorry I didn't catch this earlier: why is this returning `m, err` now? It's pretty unidiomatic for Go to return something with `err`. AFAICT the only place this is called in this PR doesn't consider the `m` return value anyway, and callers of exported `ComputeManifest` might be surprised by the behavior change.

Reply:
> It's pretty unidiomatic for Go to return something with err

I must have misunderstood you. The signature is `(m Manifest, err error)`, so you always have to return two values.

> callers of exported ComputeManifest might be surprised by the behavior change.

I still don't get it. What behavioural change? The return signature is unchanged.

Reply:
The return signature is unchanged, but the behavior is different now -- it used to return `nil, err`, and now it returns a non-nil `m`. This might surprise some callers. If there's a reason to return a potentially partial manifest in the case of an error, we should take care to document it better. Otherwise most callers will likely expect that the return is either-or.

Reply:
Oh is that what you meant? I completely misunderstood. Fixing it right now.

Reply:
Done. Fixed all the other places it does it in there as well.